text/3735-field-projections.md

```diff
@@ -0,0 +1,2022 @@
+concepts are:
+
+- [field types],
+  - `Field` traits,
+  - `field_of!` macro,
+  - [`#[projecting]`](#projecting-attribute) attribute
+- projection operator `->`,
+  - `Project` trait,
+  - `Projectable` trait,
+  - `SimultaneousProject` trait,
+  - `SimultaneousProjectable` trait,
+
+To ease understanding, here is a short explanation of the interactions between these concepts. The
+following subsections explain them in more detail, so refer to them in cases of ambiguity. The
+projection operator `->` is governed by the `Project` trait that has `Projectable` as a super trait.
+`Projectable` helps to select the struct whose fields are used for projection. Field types store
+information about a field (such as the base struct and the field type) via the `UnalignedField`
+trait and the `field_of!` macro makes it possible to name the [field types]. Finally, the
+[`#[projecting]`](#projecting-attribute) attribute allows `repr(transparent)` structs to be ignored
+when looking for fields for projection.
+
+The traits `SimultaneousProject` and `SimultaneousProjectable` exist to support simultaneous
+projections.
+
+### Field Types
+[field type]: #field-types
+[field types]: #field-types
+
+The compiler generates a compiler-internal type for every sized[^2] field of every struct and tuple.
+These types can only be named via the `field_of!` macro that has the same syntax as `offset_of!`.
+Only fields accessible to the current scope can be projected. These types are called *field types*.
+
+[^2]: This restriction can be lifted in the future to include unsized types with statically known
+    alignment, but that would have to be done in unison with adding support for those fields in
+    `offset_of!`.
+
+Field types implement the `UnalignedField` trait:
+
+```rust
+/// # Safety
+///
+/// In any instance of the type `Self::Base`, at byte offset `Self::OFFSET`, there exists a
+/// (possibly misaligned) field of type `Self::Type`.
+pub unsafe trait UnalignedField {
+    type Base: ?Sized;
+    type Type: ?Sized;
+
+    const OFFSET: usize;
+}
+```
+
+In the implementation of this trait, `Base` is set to the struct that the field is part of and
+`Type` is set to the type of the field. `OFFSET` is set to the offset in bytes of the field in
+the struct (i.e. `OFFSET = offset_of!(Base, ident)`).
+
+For aligned fields (such as all fields of non-`#[repr(packed)]` structs), their field types also
+implement the `Field` trait:
+
+```rust
+/// # Safety
+///
+/// In any well-aligned instance of the type `Self::Base`, at byte offset `Self::OFFSET`, there
+/// exists a well-aligned field of type `Self::Type`.
+pub unsafe trait Field: UnalignedField {}
+```
+
+In addition to all fields of all structs and tuples, field types are generated for transparent,
+[`#[projecting]`](#projecting-attribute) container types as follows: given a transparent, generic
+type annotated with [`#[projecting]`](#projecting-attribute) and a struct contained in it:
```

Comment:
Should this specify more strictly that only types with a single generic argument? (Which is **not** implied by transparent).

Ie, what happens to

```rust
#[projecting]
#[repr(transparent)]
struct Foo<T,U> {
    inner: T,
    yolo: PhantomData<U>
}
```

Reply:
Thanks for the question! I have updated the section, your example is allowed and only `T` will be projected.